aten/src/ATen/native/ComparisonUtils.cpp

```diff
@@ -14,23 +15,44 @@ class Tensor;
 
 namespace native {
 
-template<typename O, typename C>
-void _assert_match(const O& original, const C& compared, const std::string& name) {
-  if (compared) {
-    bool equal = (original == compared.value());
-    if (!equal) {
-      std::stringstream msg;
-      msg << "Tensor " << name << " mismatch!";
-      AT_ASSERT(equal, msg.str());
+static void _assert_equal(const std::string& name, const bool& condition) {
+    if (!condition) {
+        std::stringstream msg;
+        msg << "Tensor " << name << " mismatch!";
+        AT_ASSERT(condition, msg.str());
+    }
+}
+
+static void _assert_match(const c10::SymIntArrayRef& original, const c10::OptionalArrayRef<long int>& compared, const std::string& name) {
+    if (compared) {
+        auto originaAsIntArrayRef = c10::asIntArrayRefUnchecked(original);
+        _assert_equal(name, originaAsIntArrayRef == compared);
+    }
+}
+
+static void _assert_match(const c10::ScalarType& original, const c10::optional<c10::ScalarType>& compared, const std::string& name) {
+    if (compared) {
+        _assert_equal(name, original == compared.value());
+    }
+}
+
+static void _assert_match(const caffe2::TypeMeta& original, const c10::optional<c10::ScalarType>& compared, const std::string& name) {
+    if (compared) {
+        _assert_equal(name, original == compared);
     }
-  }
 }
 
 void _assert_tensor_metadata(at::Tensor const& tensor, at::OptionalIntArrayRef sizes, at::OptionalIntArrayRef strides, c10::optional<c10::ScalarType> dtype) {
-  _assert_match(tensor.sizes(), sizes, "sizes");
-  _assert_match(tensor.strides(), strides, "strides");
+  _assert_match(tensor.sym_sizes(), sizes, "sizes");
+  _assert_match(tensor.sym_strides(), strides, "strides");
   _assert_match(tensor.dtype(), dtype, "dtype");
 }
 
+Tensor _functional_assert_tensor_metadata(at::Tensor const& tensor, at::OptionalIntArrayRef sizes, at::OptionalIntArrayRef strides, c10::optional<c10::ScalarType> dtype) {
```

Comment:
I think, fundamentally, your real problem is that this function (and the one above) aren't taking OptionalSymIntArrayRef, they are taking OptionalIntArrayRef. From the schema below, they should accept symints. If you take symints here, then you no longer need to do conversions with asIntArrayRef, you have symints on the left and right and a straight equality will work.